src/isotp_fast.c

```diff
@@ -0,0 +1,1058 @@
+                }
+
+                if (sctx->ctx->opts->bs && !sctx->bs) {
+                    k_timer_start(&sctx->timer, K_MSEC(ISOTP_BS_TIMEOUT_MS), K_NO_WAIT);
+                    sctx->state = ISOTP_TX_WAIT_FC;
+                    LOG_DBG("BS reached. Wait for FC again");
+                    break;
+                }
+                else if (sctx->stmin) {
+                    sctx->state = ISOTP_TX_WAIT_ST;
+                    break;
+                }
+
+                /* Ensure FIFO style transmission of CF */
+                k_sem_take(&sctx->sem, K_FOREVER);
+            } while (ret > 0);
+            break;
+
+        case ISOTP_TX_WAIT_ST:
+            k_timer_start(&sctx->timer, stmin_to_timeout(sctx->stmin), K_NO_WAIT);
+            sctx->state = ISOTP_TX_SEND_CF;
+            LOG_DBG("SM wait ST");
+            break;
+
+        case ISOTP_TX_ERR:
+            LOG_DBG("SM error");
+            sctx->ctx->sent_callback(sctx->error, sctx->cb_arg);
+            sctx->state = ISOTP_TX_STATE_RESET;
+            free_send_ctx(&sctx);
+            break;
+
+            /*
+             * We sent this synchronously in isotp_fast_send.
+             * case ISOTP_TX_SEND_SF:
+             *   __fallthrough;
+             * */
+
+        case ISOTP_TX_WAIT_FIN:
+            LOG_DBG("SM finish");
+            k_timer_stop(&sctx->timer);
+
+            sctx->ctx->sent_callback(ISOTP_N_OK, sctx->cb_arg);
+            sctx->state = ISOTP_TX_STATE_RESET;
+            free_send_ctx(&sctx);
+            break;
+
+        default:
+            break;
+    }
+}
+
+static void send_work_handler(struct k_work *work)
+{
+    struct isotp_fast_send_ctx *sctx = CONTAINER_OF(work, struct isotp_fast_send_ctx, work);
+
+    send_state_machine(sctx);
+}
+
+static void send_timeout_handler(struct k_timer *timer)
+{
+    struct isotp_fast_send_ctx *sctx = CONTAINER_OF(timer, struct isotp_fast_send_ctx, timer);
+
+    if (sctx->state != ISOTP_TX_SEND_CF) {
+        LOG_ERR("Timed out waiting for FC frame");
+        send_report_error(sctx, ISOTP_N_TIMEOUT_BS);
+    }
+
+    k_work_submit(&sctx->work);
+}
+
+static inline void prepare_filter(struct can_filter *filter, isotp_fast_msg_id my_addr,
+                                  const struct isotp_fast_opts *opts)
+{
+    filter->id = my_addr;
+    filter->mask = ISOTP_FIXED_ADDR_RX_MASK;
+    filter->flags = CAN_FILTER_DATA | CAN_FILTER_IDE
+                    | ((opts->flags & ISOTP_MSG_FDF) != 0 ? CAN_FILTER_FDF : 0);
+}
+
+int isotp_fast_bind(struct isotp_fast_ctx *ctx, const struct device *can_dev,
+                    const isotp_fast_msg_id my_addr, const struct isotp_fast_opts *opts,
+                    isotp_fast_recv_callback_t recv_callback, void *recv_cb_arg,
+                    isotp_fast_recv_error_callback_t recv_error_callback,
+                    isotp_fast_send_callback_t sent_callback)
+{
+    sys_slist_init(&isotp_send_ctx_list);
+    sys_slist_init(&isotp_recv_ctx_list);
```

Comment:
Is it safe to re-initialize the `slist` with every bind? And does this still allow to have multiple binds?

Reply:
I think this did cross my mind at one point actually, but I never got round to doing anything about it. The lists should probably sit on the `isotp_fast_context` itself.